client/cache/write_delete.go

```diff
@@ -47,3 +49,39 @@ func (w *WriteDeleteCache) Set(ctx context.Context, key string, val any) error {
 	}
 	return w.Cache.Delete(ctx, key)
 }
+
+// WriteDoubleDeleteCache creates write double delete cache pattern decorator.
+// The fn is the function that persistent the key and val.
+type WriteDoubleDeleteCache struct {
+	Cache
+	interval  time.Duration
+	storeFunc func(ctx context.Context, key string, val any) error
+}
+
+func NewWriteDoubleDeleteCache(cache Cache, interval time.Duration,
+	fn func(ctx context.Context, key string, val any) error) (*WriteDoubleDeleteCache, error) {
+	if fn == nil || cache == nil {
+		return nil, berror.Error(InvalidInitParameters, "cache or storeFunc can not be nil")
+	}
+
+	return &WriteDoubleDeleteCache{
+		Cache:     cache,
+		interval:  interval,
+		storeFunc: fn,
+	}, nil
+}
+
+func (c *WriteDoubleDeleteCache) Set(ctx context.Context, key string, val any) error {
+	err := c.storeFunc(ctx, key, val)
+	if err != nil && !errors.Is(err, context.DeadlineExceeded) {
+		return berror.Wrap(err, PersistCacheFailed, fmt.Sprintf("key: %s, val: %v", key, val))
+	}
+	go func() {
+		timer := time.NewTimer(c.interval)
+		defer timer.Stop()
+		<-timer.C
+		_ = c.Cache.Delete(ctx, key)
```

Comment:
What about using the time.AfterFunc here?
and when you call the c.Cache.Delete, you can not use the ctx and please create a new one with 1s expiration.